### ISSUE TITLE
What should `local_accessor().max_size()` return?

### ISSUE THREAD

**Issue by maarquitos14** [KhronosGroup/SYCL-Docs#332]
On the one hand, according to Table 77, `local_accessor()` 
```
Constructs an empty local accessor which fulfills the following post-conditions:
- (empty() == true)
- All size queries return 0.
- The return values of get_pointer() and get_multi_ptr() are unspecified.
- Trying to access the underlying memory is undefined behavior.
```

On the other hand, according to Table 80, `max_size()` 
```
Returns the maximum number of elements any accessor of this type would be able to access.
```

So, should `local_accessor().max_size()` return 0, following Table 77 (considering that `max_size()` is one of `All size queries`), or should it return `the maximum number of elements any accessor of this type would be able to access` following Table 80?

For the record, _my_ understanding is that `All size queries` refer to the queries related to the instance, while `max_size()` is a generic query not actually related to the instance, so `max_size()` should not return 0.

**Comment by gmlueck** [comment]
This is a interesting question.  We added `max_size` in order to align with the C++ requirements for "Container".  When looking at standard C++ containers, I see that the definition of `std::array` says this about `max_size`:

> Because each `std::array<T, N>` is a fixed-size container, the value returned by `max_size` equals `N` (which is also the value returned by [`size`](https://en.cppreference.com/w/cpp/container/array/size))

(https://en.cppreference.com/w/cpp/container/array/max_size)

Note that `max_size` is not a static member function, so it applies to a specific instance of the container.  If the intent was to return the maximum size of any possible instance, I presume it would be a static member function, but that is not the case.

Since all accessor instances have a fixed size after construction, I think `max_size` should return the same value as `size`.  Opinions?

**Comment by tomdeakin** [comment]
The description of the function suggests to me that it should be a static member function and so return the maximum size of any local accessor could be. IMO, we should clarify that this should be `static`.

**Comment by gmlueck** [comment]
But the `max_size` function is defined by standard C++ Container classes, and it is not static there.  SYCL should match standard C++ here, I think.

**Comment by keryell** [comment]
And if we have a placeholder accessor, I guess in that case we should return a huge value, similar to the proposed static member function.
But there is no local placeholder accessor...

**Comment by gmlueck** [comment]
Correct, there is no issue here for placeholder accessors.  A placeholder buffer accessor still has a fixed size after construction, and there is no such thing as a local placeholder accessor.

**Comment by Pennycook** [comment]
> This is a interesting question. We added `max_size` in order to align with the C++ requirements for "Container".

Does anybody remember why we did this?

The first thing that ISO C++ says about containers is [[container.requirements]](https://eel.is/c++draft/containers#container.requirements.pre-1):
> Containers are objects that store other objects[.](https://eel.is/c++draft/containers#container.requirements.pre-1.sentence-1) They control allocation and deallocation of these objects through constructors, destructors, insert and erase operations[.](https://eel.is/c++draft/containers#container.requirements.pre-1.sentence-2)

I don't think this describes `accessor` at all.  I think an `accessor` is a _view_ of data managed by something else (i.e., a `buffer`).

Pretending `accessor` is a container doesn't seem consistent with the rest of the specification, because it implies that two different `accessor` objects created from the same `buffer` would store independent copies of the `buffer` contents, and I don't think this is necessarily true for all implementations. 

Would it make more sense to try and align the `accessor` interface with [[ranges]](https://eel.is/c++draft/ranges) instead?  If `accessor`s were "container-like" (this is an ISO C++ term) and implemented the [[view.interface]](https://eel.is/c++draft/view.interface), then I think it would be sufficient to expose:
- `begin()` and `end()`
- `data()`
- `size()`
- `empty()`

We can keep `cbegin`, `cend`, `rbegin` and `rend` since they're already there.

I think what this would amount to is the following:
- Deprecate `max_size()`: it doesn't appear to be part of the `ranges` interface, and doesn't make sense for `accessor`.
- Add `accessor::data()`, defined as returning `get_pointer() + get_offset()` (i.e., pointing at the beginning of the range).  This is aligned with `ranges` (where `begin()` and `data()` are often equivalent), and solves the longstanding problem of `get_pointer()` being weird.
